iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,435 @@
+
+        Parameters
+        ----------
+        ctf_params : list of dicts
+            Each dict contains CTF k,v pairs per particle.
+            Shape (n_particles,)
+
+        Returns
+        -------
+        ctfs : arr
+            Shape (n_ctfs, n_pix, n_pix)
+        """
+        n_ctfs = len(ctf_params)
+        ctfs = []
+
+        for i in range(n_ctfs):
+            ctfs.append(eval_ctf(**ctf_params[i]))
+
+        return ctfs
+
+    ctfs = do_build_ctf(ctf_info)
+    ctfs_1, ctfs_2 = do_split(ctfs)
+
+    # work in Fourier space. as particles stay in Fourier space the whole time.
+    # they are experimental measurements and are fixed in the algorithm
+    particles_1, particles_2 = do_split(particles)
+    particles_f_1 = do_fft(particles_1, d=3)
+    particles_f_2 = do_fft(particles_2, d=3)
+
+    n_pix = map_3d_init.shape[0]
+
+    max_n_iters = 7  # in practice programs using 3-10 iterations.
+
+    half_map_3d_r_1, half_map_3d_r_2 = map_3d_init, map_3d_init.copy()
+
+    for _ in range(max_n_iters):
+
+        # TODO: implement 3D fft using torch
+        half_map_3d_f_1 = do_fft(half_map_3d_r_1, d=3)
+        half_map_3d_f_2 = do_fft(half_map_3d_r_2, d=3)
+
+        def grid_SO3_uniform(n_rotations):
+            """
+            Generate a discrete set of uniformly distributed rotations
+            across SO(3).
+
+            Parameters
+            ----------
+            n_rotations : int
+                Number of rotations
+
+            Returns
+            -------
+            rots : arr
+                List of rotations.
+                Shape (n_rotations, 3, 3)
+            """
+            rots = np.ones((n_rotations, 3, 3))
+            return rots
+
+        n_rotations = 1000
+        rots = grid_SO3_uniform(n_rotations)
+
+        def generate_xy_plane(n_pix):
+            """
+            Generate xy plane.
+
+            Parameters
+            ----------
+            n_pix : int
+                Number of pixels
+
+            Returns
+            xy_plane : arr
+                Array describing xy plane in space.
+                Shape (n_pix**2, 3)
+            """
+
+            # See how meshgrid and generate coordinates functions used
+            # TODO:
+            # https://github.com/geoffwoollard/compSPI/blob/stash_simulate/src/simulate.py#L96
+
+            xy_plane = np.ones((n_pix**2, 3))
+            return xy_plane
+
+        xy_plane = generate_xy_plane(n_pix)
+
+        def do_slices(map_3d_f, rots):
+            """
+            Generates slice coordinates by rotating xy plane.
+            Interpolate values from map_3d_f onto 3D coordinates.
+            TODO: See how scipy map_values used to interpolate in
+            https://github.com/geoffwoollard/compSPI/blob/stash_simulate/src/simulate.py#L111
+
+            Parameters
+            ----------
+            map_3d_f : arr
+                Shape (n_pix, n_pix, n_pix)
+            rots : arr
+                List of rotations.
```

Comment:
These should be arrays so you can do vectorized operations. Not a list of arrays.

Reply:
`List of rotations` makes it sound like its a list.